### PR TITLE
[1769] Order flow for ISS

### PR DESCRIPTION
**Pull request by tonyheadford**
### Context
[Trello card](https://trello.com/c/xppMclNt/1769-order-flow-for-iss)
Order flow for state-funded pupils in independent settings

### Changes proposed in this pull request
Added pages for device ordering flow for `LaFundedPlace` type `School`

### Guidance to review
__Setup__
1. Add a `LaFundedPlace` school to a LocalAuthority e.g.
```ruby
my_la_funded_place = LocalAuthority.first.create_la_funded_places!(urn: 999000, device_allocation: 10)
```
2. Add a user to the school
```ruby
my_la_funded_place.users << my_user
```
3. Up the device cap to the allocation amount
```ruby
my_la_funded_place.std_device_allocation.update!(cap: 10)
```
4. Set the new school to order 
```ruby
my_la_funded_place.can_order!
```
5. Sign-in as the user associated with the school (and choose the school after sign-in if necessary)
6. The journey starts at the home page for the school and by following the `Get laptops` link
image
image
image
image
image
image
image

### CODE REVIEW

File: spec/form_objects/chromebook_information_form_spec.rb [hunk 1]
```diff
@@ -56,4 +56,27 @@
...
+    context 'when :will_need_chromebooks is blank' do
+      let(:will_need_chromebooks) { nil }
+
+      it 'is not valid' do
+        expect(form.valid?).to be_falsey
+      end
+
+      it 'has an error message including the schools institution_type' do
+        form.validate
+        # TODO: update message
```
Comment (nigel-lowry) [inline spec/form_objects/chromebook_information_form_spec.rb:77]: Still got a TODO here

File: spec/models/preorder_information_spec.rb [hunk 2]
```diff
@@ -372,4 +372,45 @@
...
+      context 'when chromebooks needed' do
+        before do
+          preorder_info.update!(will_need_chromebooks: 'yes',
+                                school_or_rb_domain: 'school.org',
+                                recovery_email_address: 'school@example.com')
+        end
+
+        it 'does not require a G Suite domain' do
+          preorder_info.school_or_rb_domain = nil
+          expect(preorder_info.chromebook_information_complete?).to be true
```
Comment (nigel-lowry) [inline spec/models/preorder_information_spec.rb:392]: Another way of writing this you may prefer is: `expect(preorder_info).to be chromebook_information_complete`. If so there's a few occurrences in this file

File: spec/form_objects/chromebook_information_form_spec.rb (hunk 1 above)
Comment (tonyheadford) [reply spec/form_objects/chromebook_information_form_spec.rb]: Good spot

File: spec/models/preorder_information_spec.rb (hunk 2 above)
Comment (tonyheadford) [reply spec/models/preorder_information_spec.rb:392]: My eyes!

File: app/controllers/school/la_funded_places_chromebooks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  end
+
+  def chromebook_params
+    params.fetch(:chromebook_information_form, {}).permit(
+      :will_need_chromebooks,
+    )
+  end
+
+  def custom_error_message
+    'Tell us if pupils will need Chromebooks'
```
Comment (jamesrose) [inline app/controllers/school/la_funded_places_chromebooks_controller.rb:43]: This might be better placed in `config/locales/en.yml`

File: app/form_objects/chromebook_information_form.rb [hunk 4]
```diff
@@ -36,4 +37,25 @@ def recovery_email_address_cannot_be_same_domain_as_school_or_rb
...
     end
   end
+
+  def will_need_chromebooks_and_is_not_a_la_funded_school?
+    if will_need_chromebooks?
+      if school.respond_to?(:la_funded_place?)
+        !school.la_funded_place?
+      else
+        true
+      end
```
Comment (jamesrose) [inline app/form_objects/chromebook_information_form.rb:47]:
```suggestion
     !school.la_funded_place?
```

Do we still need to check if school responds to this method?

File: app/models/la_funded_place.rb [hunk 5]
```diff
@@ -1,5 +1,9 @@
 class LaFundedPlace < CompulsorySchool
   def institution_type
-    'funded_places'
+    'local authority'
```
Comment (jamesrose) [inline app/models/la_funded_place.rb:3]:
```suggestion
    'local_authority'
```

I'm pretty sure this should be underscored.

File: app/models/la_funded_place.rb [hunk 6]
```diff
@@ -1,5 +1,9 @@
...
+  end
+
+  def techsource_urn
+    "iss #{responsible_body.name.downcase}"
```
Comment (jamesrose) [inline app/models/la_funded_place.rb:7]:
```suggestion
    "iss#{responsible_body.name.downcase}"
```

Probably safer to not have a space here as it could impact CC downstream.

File: app/models/preorder_information.rb [hunk 7]
```diff
@@ -109,7 +109,14 @@ def will_need_chromebooks?
     will_need_chromebooks == 'yes'
   end
 
+  def will_not_need_chromebooks?
+    will_need_chromebooks == 'no'
+  end
+
   def chromebook_information_complete?
+    # do not want to hold them up being able to order
+    return true if school&.la_funded_place_establishment_type?
```
Comment (jamesrose) [inline app/models/preorder_information.rb:118]:
```suggestion
    return true if school.la_funded_place_establishment_type?
```

We can safely assume that there will be a school on this model.

File: app/models/school_device_allocation.rb [hunk 8]
```diff
@@ -88,6 +88,14 @@ def raw_allocation
     self[:allocation]
   end
 
+  def devices_available_to_order
+    if has_virtual_cap_feature_flags? && is_in_virtual_cap_pool?
+      cap - devices_ordered
+    else
+      raw_cap - raw_devices_ordered
+    end
```
Comment (jamesrose) [inline app/models/school_device_allocation.rb:96]:
```suggestion
    cap - devices_ordered
```

Won't `cap` and `devices_ordered` do this themselves?

File: app/views/school/la_funded_places/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    <p class="govuk-body">
+      <%= govuk_link_to 'Check what types of laptop are available', laptop_types_school_path(@school) %>
+    </p>
+
+    <h2 class="govuk-heading-m govuk-!-margin-top-6">
+      Ownership and responsibility
+    </h2>
+    <p class="govuk-body">
+    <%= @school.responsible_body.name %> will own these laptops and is responsible for:
+    </p>
```
Comment (jamesrose) [inline app/views/school/la_funded_places/show.html.erb:67]:
```suggestion
    <p class="govuk-body">
      <%= @school.responsible_body.name %> will own these laptops and is responsible for:
    </p>
```

nit: Please could you look at the indentation of this file?

File: config/locales/en.yml [hunk 10]
```diff
@@ -62,6 +62,11 @@ en:
     general_privacy_notice: How we look after personal information as part of the Get help with technology programme
     what_to_do_if_you_cannot_get_laptops_tablets_or_internet_access_from_dfe: What to do if you cannot get laptops, tablets or internet access from DfE
     how_to_access_the_get_help_with_technology_service: How to access the Get help with technology service
+    iss_home: Get laptops and internet access
+    iss_order_devices: Get laptops
+    iss_details: Check your details
+    iss_laptop_types: Find out what type of laptop pupils need
+    iss_chromebooks: Will your order include Google Chromebooks?
```
Comment (jamesrose) [inline config/locales/en.yml:69]:
```suggestion
    iss:
      home: Get laptops and internet access
      order_devices: Get laptops
      details: Check your details
      laptop_types: Find out what type of laptop pupils need
      chromebooks: Will your order include Google Chromebooks?
```

I'd scope these instead of prefixing.

**Review by jamesrose — COMMENTED** [review]
Looking great. Just a few comments, but nothing ground-breaking. I didn't review the content or design itself as part of this review. I'll wait until Kate has reviewed to do that.

File: app/models/la_funded_place.rb (hunk 6 above)
Comment (tonyheadford) [reply app/models/la_funded_place.rb:7]: Hmm yes, although the RB names might have spaces too, maybe we need something more robust. Also I'm not sure this has been agreed with CC yet?

File: app/models/preorder_information.rb (hunk 7 above)
Comment (tonyheadford) [reply app/models/preorder_information.rb:118]: There was a whole bunch of model specs that failed because they weren't properly set up with a school. Perhaps I shouldn't be lazy and fix all the specs!

File: app/models/la_funded_place.rb (hunk 5 above)
Comment (tonyheadford) [reply app/models/la_funded_place.rb:3]: Good spot

File: app/models/la_funded_place.rb (hunk 6 above)
Comment (jamesrose) [reply app/models/la_funded_place.rb:7]: No, I think we would need to check what they'd like in this instance @markhallen

File: app/views/school/la_funded_places/order.html.erb [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    </p>
+    <p class="govuk-body">
+    <% devices_available = @school.std_device_allocation.devices_available_to_order %>
+    If you place an order for more than <%= "#{devices_available} #{'laptop'.pluralize(devices_available)}" %>, your order will be cancelled.
+    </p>
+    <p class="govuk-body">
+      You can order all of your laptops at once, or place multiple orders.
+    </p>
+
+    <%# instead of partial shared/start_ordering_on_techsource %>
```
Comment (asmega) [inline app/views/school/la_funded_places/order.html.erb:102]: delete me?

File: app/models/local_authority.rb [hunk 12]
```diff
@@ -36,10 +36,10 @@ def create_la_funded_places!(urn:, device_allocation: 0, router_allocation: 0, e
       postcode: postcode,
     }.reverse_merge(extra_args)
 
-    funded_places = LaFundedPlace.create!(attrs)
-    funded_places.create_preorder_information!(who_will_order_devices: 'responsible_body', will_need_chromebooks: 'no')
-    funded_places.device_allocations.std_device.create!(allocation: device_allocation)
-    funded_places.device_allocations.coms_device.create!(allocation: router_allocation)
-    funded_places
+    funded_place = LaFundedPlace.create!(attrs)
+    funded_place.create_preorder_information!(who_will_order_devices: 'responsible_body')
```
Comment (asmega) [inline app/models/local_authority.rb:40]: not part of this change but are schools not ordering? otherwise we have to attach users to RB who will have access to everything?

File: app/models/school_device_allocation.rb (hunk 8 above)
Comment (tonyheadford) [reply app/models/school_device_allocation.rb:96]: Doh!

File: app/views/school/home/show_la_funded_place.html.erb [hunk 13]
```diff
@@ -0,0 +1,60 @@
+<%- title = t('page_titles.iss_home') %>
```
Comment (asmega) [inline app/views/school/home/show_la_funded_place.html.erb:1]: this translation is not defined and therefore missing

Comment (asmega) [reply app/views/school/home/show_la_funded_place.html.erb:1]: looks a review commit in haste has caused issues

File: app/models/local_authority.rb (hunk 12 above)
Comment (tonyheadford) [reply app/models/local_authority.rb:40]: Yes, it should be schools now, I think originally we thought it would be the RB, I'll change that and make sure it all still works 😬

File: app/views/school/home/show_la_funded_place.html.erb (hunk 13 above)
Comment (tonyheadford) [reply app/views/school/home/show_la_funded_place.html.erb:1]: Good spot, that's because I accepted James change  in the comment to scope the `iss_` names

File: app/models/la_funded_place.rb (hunk 6 above)
Comment (asmega) [reply app/models/la_funded_place.rb:7]: Also this method is only used in a view to playback to a user. Does it not need to hook in somewhere where we need to provide this info to CC so they can correlate things?

Comment (tonyheadford) [reply app/models/la_funded_place.rb:7]: Yes it will have to at some point

Comment (tonyheadford) [reply app/models/la_funded_place.rb:7]: Just fyi - @ralph-hawkins had highlighted that we need to differentiate between ISS and Social Care and had suggested the __iss-__ and __slc-__ prefix

File: app/models/preorder_information.rb (hunk 7 above)
Comment (tonyheadford) [reply app/models/preorder_information.rb:118]: On further investigation I've left it in. The `chromebook_information_complete?` method gets called from `infer_status` which itself gets called via callbacks during the creation of objects in specs. Removing the `&` causes 400+ specs to fail.

File: app/views/school/la_funded_places/show.html.erb (hunk 9 above)
Comment (tonyheadford) [reply app/views/school/la_funded_places/show.html.erb]: Done

File: app/views/school/la_funded_places/order.html.erb (hunk 11 above)
Comment (tonyheadford) [reply app/views/school/la_funded_places/order.html.erb]: Done

File: app/views/school/home/show_la_funded_place.html.erb (hunk 13 above)
Comment (tonyheadford) [reply app/views/school/home/show_la_funded_place.html.erb]: Done

File: app/models/school_device_allocation.rb (hunk 8 above)
Comment (tonyheadford) [reply app/models/school_device_allocation.rb]: Done

File: app/form_objects/chromebook_information_form.rb (hunk 4 above)
Comment (tonyheadford) [reply app/form_objects/chromebook_information_form.rb]: Good call, we don't need this now, I've remove it.

File: app/controllers/school/la_funded_places_chromebooks_controller.rb (hunk 3 above)
Comment (tonyheadford) [reply app/controllers/school/la_funded_places_chromebooks_controller.rb]: Yes, done.

File: app/models/local_authority.rb (hunk 12 above)
Comment (tonyheadford) [reply app/models/local_authority.rb]: Change made and seems to work fine when set up as `school` so far.

File: app/views/school/home/show_la_funded_place.html.erb [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds-from-desktop govuk-grid-column-full">
+    <h2 class="govuk-heading-l govuk-!-font-size-27">
+      <%= govuk_link_to t('page_titles.iss.order_devices'), order_devices_school_path(@school) %>
+    </h2>
+
+    <p class="govuk-body">Use this section to:</p>
+    <ul class="govuk-list govuk-list--bullet">
+      <li>see how many laptops you have been allocated</li>
```
Comment (Lloyd-K) [inline app/views/school/home/show_la_funded_place.html.erb:24]: see how many laptops you've been allocated

File: app/views/school/la_funded_places/laptop_types.html.erb [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+      DfE Restricted Windows laptops
+    </h2>
+    <p class="govuk-body">
+      These have temporary safeguarding settings installed — content filtering and mobile device management software — which you’ll need to replace when they expire on 30 September 2021. You’ll need to reset these devices before you can apply new settings, or download some types of software.
+    </p>
+    <h2 class="govuk-heading-s">
+      Google Chromebooks
+    </h2>
+    <p class="govuk-body">
+      You’ll need to provide us with the relevant school’s email domain for <span class="app-no-wrap">G Suite for Education</span>, and a recovery email address.
```
Comment (Lloyd-K) [inline app/views/school/la_funded_places/laptop_types.html.erb:34]: ... email domain for Google Workspace for Education Fundamentals (formerly G Suite for Education), and a recovery ...

File: app/views/school/la_funded_places/laptop_types.html.erb [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+  <%- school_breadcrumbs items: [ { t('page_titles.iss.order_devices') => order_devices_school_path(@school)}, title], school: @school, user: impersonated_or_current_user %>
+<%- end %>
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+    <h1 class="govuk-heading-xl">
+      <%= title %>
+    </h1>
+    <p class="govuk-body">
+      We recommend checking with schools, families or pupils as soon as possible to find out which device type and IT settings would be most suitable.
```
Comment (Lloyd-K) [inline app/views/school/la_funded_places/laptop_types.html.erb:13]: ... checking with your IT team, schools, families ...

File: app/views/school/la_funded_places/order.html.erb [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+          Change <span class="govuk-visually-hidden">Chromebook preference</span>
+        <% end %>
+      </p>
+      <div class="govuk-inset-text">
+        <p class="govuk-body">
+          After you place your order you’ll get an email asking for:
+        </p>
+        <ul class="govuk-list govuk-list--bullet">
+          <li>
+            the domain the school has registered for G Suite for Education, for example ‘school.co.uk’
```
Comment (Lloyd-K) [inline app/views/school/la_funded_places/order.html.erb:49]: ... for Google Workspace for Education Fundamentals (formerly G Suite for Education), for example ...

File: app/views/school/la_funded_places/order.html.erb [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+          </li>
+          <li>
+            a recovery email address. This should be on a separate domain to the one registered for G Suite for Education. For example, if the domain registered for G Suite for Education is ‘school.com’, the recovery email cannot be ‘recovery@school.com’
```
Comment (Lloyd-K) [inline app/views/school/la_funded_places/order.html.erb:52]: ... registered for Google Workspace for Education. For example ...

File: app/views/school/la_funded_places/order.html.erb [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+      </p>
+    <% else %>
+      <div class="govuk-inset-text">
+        <p class="govuk-body">
+          <strong>
+            You have not told us if you plan to order Google Chromebooks
+          </strong>
+        </p>
+        <p class="govuk-body">
+          If you order Chromebooks without letting us know first, they’ll be delivered without safeguarding settings and access to G Suite for Education.
```
Comment (Lloyd-K) [inline app/views/school/la_funded_places/order.html.erb:74]: ... delivered unsecured, without access to Google Workspace for Education Fundamentals (formerly G Suite for Education).

File: app/views/school/la_funded_places/show.html.erb [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+    <%= govuk_details summary: 'How your allocation has been calculated' do %>
+      <p class="govuk-body">
+        Your allocation estimate is based on the data <%= @school.responsible_body.name %>
+        submitted in the latest published alternative provision census (2020).
+      </p>
+      <p class="govuk-body">
+        We’ve allocated one laptop for each pupil in years 3 to 11 that
+        <%= @school.responsible_body.name %> reported as having a state-funded
+        place at an independent setting, and also being eligible for free school meals.
```
Comment (Lloyd-K) [inline app/views/school/la_funded_places/show.html.erb:27]: ... at an independent special school or alternative provision, and also ..

File: app/views/school/la_funded_places/show.html.erb [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+        Do not send any personal information, like names of pupils or dates of birth.
+      </p>
+      <p class="govuk-body">
+        Email this information to <%= ghwt_contact_mailto %>.
+      </p>
+    <% end %>
+
+    <h2 class="govuk-heading-m govuk-!-margin-top-6">Before you order</h2>
+    <p class="govuk-body">
+      We recommend checking with schools, families or pupils as soon as possible
```
Comment (Lloyd-K) [inline app/views/school/la_funded_places/show.html.erb:57]: ... with your IT team, schools, families ...

File: app/views/school/la_funded_places_chromebooks/edit.html.erb [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+      <p class="govuk-body">
+        We need to know this so we can apply the right settings to the laptops you order. You can change your answer later.
+      </p>
+
+      <%= f.govuk_radio_buttons_fieldset( :will_need_chromebooks, legend: {text: ""} ) do %>
+        <%= f.govuk_radio_button(:will_need_chromebooks,
+                                 'yes',
+                                 label: { text: t(:yes, scope: scope) }) do %>
+          <p class="govuk-body">
+            When you order you’ll be contacted by email for G Suite for Education license information.
```
Comment (Lloyd-K) [inline app/views/school/la_funded_places_chromebooks/edit.html.erb:27]: ... you'll receive an email asking you to provide relevant Google Workspace for Education Fundamentals (formerly G Suite for Education) license information.

File: app/views/school/la_funded_places_chromebooks/edit.html.erb [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+          </p>
+        <%- end %>
+        <%= f.govuk_radio_button  :will_need_chromebooks,
+                                  'no',
+                                  label: { text: t(:no, scope: scope) } %>
+        <%= f.govuk_radio_button(:will_need_chromebooks,
+                                 'i_dont_know',
+                                 label: { text: t(:im_not_sure, scope: scope) }) do %>
+          <p class="govuk-body">
+            If you order Chromebooks without letting us know first, they’ll be delivered without safeguarding settings and access to G Suite for Education.
```
Comment (Lloyd-K) [inline app/views/school/la_funded_places_chromebooks/edit.html.erb:37]: ... delivered unsecured, without access to Google Workspace for Education Fundamentals (formerly G Suite for Education).

File: config/locales/en.yml [hunk 24]
```diff
@@ -703,8 +709,10 @@ en:
           predecessor_split_school: Predecessor - Split School
       preorder_information:
         will_need_chromebooks:
-          "yes": Yes, we will order Chromebooks
+          'yes': Yes, we’ll order Chromebooks
           'no': No, we will not order Chromebooks
+          i_dont_know: I don’t know
```
Comment (Lloyd-K) [inline config/locales/en.yml:714]: I do not know

**Review by Lloyd-K — CHANGES_REQUESTED** [review]
Please see comments.

File: app/views/school/home/show_la_funded_place.html.erb (hunk 14 above)
Comment (tonyheadford) [reply app/views/school/home/show_la_funded_place.html.erb:24]: Done

File: app/views/school/la_funded_places/laptop_types.html.erb (hunk 15 above)
Comment (tonyheadford) [reply app/views/school/la_funded_places/laptop_types.html.erb:34]: Done

File: app/views/school/la_funded_places/laptop_types.html.erb (hunk 16 above)
Comment (tonyheadford) [reply app/views/school/la_funded_places/laptop_types.html.erb:13]: Done

File: app/views/school/la_funded_places/order.html.erb (hunk 17 above)
Comment (tonyheadford) [reply app/views/school/la_funded_places/order.html.erb:49]: Done

File: app/views/school/la_funded_places/order.html.erb (hunk 18 above)
Comment (tonyheadford) [reply app/views/school/la_funded_places/order.html.erb:52]: Done

File: app/views/school/la_funded_places/order.html.erb (hunk 19 above)
Comment (tonyheadford) [reply app/views/school/la_funded_places/order.html.erb:74]: Done

File: app/views/school/la_funded_places/show.html.erb (hunk 20 above)
Comment (tonyheadford) [reply app/views/school/la_funded_places/show.html.erb:27]: Done

File: app/views/school/la_funded_places/show.html.erb (hunk 21 above)
Comment (tonyheadford) [reply app/views/school/la_funded_places/show.html.erb:57]: Done

File: app/views/school/la_funded_places_chromebooks/edit.html.erb (hunk 22 above)
Comment (tonyheadford) [reply app/views/school/la_funded_places_chromebooks/edit.html.erb:27]: Done

File: app/views/school/la_funded_places_chromebooks/edit.html.erb (hunk 23 above)
Comment (tonyheadford) [reply app/views/school/la_funded_places_chromebooks/edit.html.erb:37]: Done

File: config/locales/en.yml (hunk 24 above)
Comment (tonyheadford) [reply config/locales/en.yml:714]: Done

**Comment by tonyheadford** [comment]
> Please see comments.

All done

File: app/views/school/la_funded_places/laptop_types.html.erb [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+      Standard Windows laptops
+    </h2>
+    <p class="govuk-body">
+      You’ll be responsible for arranging and maintaining safeguarding measures, like content filtering.
+    </p>
+    <h2 class="govuk-heading-s">
+      DfE Restricted Windows laptops
+    </h2>
+    <p class="govuk-body">
+      These have temporary safeguarding settings installed — content filtering and mobile device management software — which you’ll need to replace when they expire on 30 September 2021. You’ll need to reset these devices before you can apply new settings, or download some types of software.
```
Comment (Lloyd-K) [inline app/views/school/la_funded_places/laptop_types.html.erb:28]: ... replace before they expire ...

**Review by Lloyd-K — CHANGES_REQUESTED** [review]
Thanks for applying the changes, I've added one extra that I missed earlier. After that, I'll approve. Thanks

File: app/views/school/la_funded_places/laptop_types.html.erb (hunk 25 above)
Comment (tonyheadford) [reply app/views/school/la_funded_places/laptop_types.html.erb]: Done

**Review by Lloyd-K — APPROVED** [review]
